tracing: Only prepare tracepoint arguments when actually tracing

0xB10C opened the pull request:
Currently, if the tracepoints are compiled (e.g. in depends and release builds), we always prepare the tracepoint arguments regardless of the tracepoints being used or not. We made sure that the argument preparation is as cheap as possible, but we can almost completely eliminate any overhead for users not interested in the tracepoints (the vast majority), by gating the tracepoint argument preparation with an `if(something is attached to this tracepoint)`. To achieve this, we use the optional semaphore feature provided by SystemTap.

The first commit simplifies and deduplicates our tracepoint macros from 13 TRACEx macros to a single TRACEPOINT macro. This makes them easier to use and also avoids more duplicate macro definitions in the second commit.

The Linux tracing tools I'm aware of (bcc, bpftrace, libbpf, and systemtap) all support the semaphore gating feature. Thus, all existing tracepoints and their argument preparation is gated in the second commit. For details, please refer to the commit messages and the updated documentation in `doc/tracing.md`.

Also adding unit tests that include all tracepoint macros to make sure there are no compiler problems with them (e.g. some varadiac extension not supported). 

Reviewers might want to check:
- Do the tracepoints still work for you? Do the examples in `contrib/tracing/` run on your system (as bpftrace frequently breaks on every new version, please test master too if it should't work for you)? Do the CI interface tests still pass?
- Is the new documentation clear? 
- The `TRACEPOINT_SEMAPHORE(event, context)` macros places global variables in our global namespace. Is this something we strictly want to avoid or maybe move to all `TRACEPOINT_SEMAPHORE`s to a separate .cpp file or even namespace? I like having the `TRACEPOINT_SEMAPHORE()` in same file as the `TRACEPOINT()`, but open for suggestion on alternative approaches. 
- Are newly added tracepoints in the unit tests visible when using `readelf -n build/src/test/test_bitcoin`? You can run the new unit tests with `./build/src/test/test_bitcoin --run_test=util_trace_tests* --log_level=all`.
<details><summary>Two of the added unit tests demonstrate that we are only processing the tracepoint arguments when attached by having a test-failure condition in the tracepoint argument preparation. The following bpftrace script can be used to demonstrate that the tests do indeed fail when attached to the tracepoints.</summary>

`fail_tests.bt`:

```c
#!/usr/bin/env bpftrace

usdt:./build/src/test/test_bitcoin:test:check_if_attached {
  printf("the 'check_if_attached' test should have failed\n");
}

usdt:./build/src/test/test_bitcoin:test:expensive_section {
  printf("the 'expensive_section' test should have failed\n");
}
```

Run the unit tests with `./build/src/test/test_bitcoin` and start `bpftrace fail_tests.bt -p $(pidof test_bitcoin)` in a separate terminal. The unit tests should fail with:

```
Running 594 test cases...
test/util_trace_tests.cpp(31): error: in "util_trace_tests/test_tracepoint_check_if_attached": check false has failed
test/util_trace_tests.cpp(51): error: in "util_trace_tests/test_tracepoint_manual_tracepoint_active_check": check false has failed

*** 2 failures are detected in the test module "Bitcoin Core Test Suite"
```

</details>

These links might provide more contextual information for reviewers:
- [How SystemTap Userspace Probes Work by eklitzke](https://eklitzke.org/how-sytemtap-userspace-probes-work) (actually an example on Bitcoin Core; mentions that with semaphores "the overhead for an untraced process is effectively zero.")
- [libbpf comment on USDT semaphore handling](https://github.com/libbpf/libbpf/blob/1596a09b5de2a50ab8d44218fc29b6d42f886305/src/usdt.c#L83-L92) (can recommend the whole comment for background on how the tracepoints and tracing tools work together)
- https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation#Semaphore_Handling

luke-jr commented on src/util/trace.h:
```diff
@@ -7,39 +7,58 @@
+//      TRACEPOINT_UNCHECKED(context, event, result);
+//  }
+//
+#define TRACEPOINT_UNCHECKED(context, event, args...) \
```
Aren't variadic macros (at least in this form) an extension, not standard C++?

0xB10C replied:
TIL, could change to `...` and `__VA_ARGS__` if needed

0xB10C commented on src/util/trace.h:
```diff
@@ -11,16 +11,54 @@
+    if (TRACEPOINT_ACTIVE(context, event)) \
+        STAP_PROBEV(context, event, args)
+
+// A USDT tracepoint with zero to twelve arguments. Here, it is **NOT** checked
+// that the tracepoint is active before preparing its arguments. Use this macro
+// if you want to, for example, prepare expensive tracepoint arguments over
+// multiple lines. Example usage:
+//
+//  if (TRACEPOINT_ACTIVE(context, event)) {
+//      result = slightly_expensive_calulation();
+//      TRACEPOINT_UNCHECKED(context, event, result);
+//  }
+//
+#define TRACEPOINT_UNCHECKED(context, event, args...) \
```
Thinking about it: We might not need an extra `TRACEPOINT_UNCHECKED` macro. We can also use the `TRACEPOINT` in all places where we'd use the unchecked one. This would mean would do two `TRACEPOINT_ACTIVE` checks, but these are extremely cheap and might even be optimized to one by the compiler. 

Dropping the unchecked version results in is a smaller, better internal tracepoint API and removes the possibility that someone uses the unchecked but doesn't gate it.

0xB10C commented:
- Addressed https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1034160141 by using `__VA_ARGS__`. Also using a separate TRACEPOINT0 macro for tracepoints without arguments, as using the TRACEPOINT macro without arguments is only be possible with GCC (clang warns that were using gcc extensions). Tested with both clang and gcc.
- Adressed https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1034548969 by dropping the `TRACEPOINT_UNCHECKED` macro.
- Added unit tests for the TRACEPOINT macros. The tracepoints in the unit tests don't do anything in particular, but the tests show that the macros work. Through the CI we cover `ENABLE_TRACING` being defined and undefined. The test case `test_tracepoint_check_if_attached` also shows that the `TRACEPOINT_ACTIVE` macro is not broken (it's tested that it's working in the functional tests).

jb55 commented:
genius. Concept ACK

vasild commented:
Concept ACK

kouloumos commented:
Concept ACK!
I've though about something similar while working on https://github.com/bitcoin/bitcoin/pull/25541 and even more when you mentioned performance impact on https://github.com/bitcoin/bitcoin/pull/25832.

0xB10C committed:
tracing: use bitcoind pid in bcc tracing examples

BCC needs the PID of a bitcoind process to attach to the tracepoints
(instead of the binary path used before) when the tracepoints have a
semaphore.

For reference, we already use the PID in our tracepoint interface
tests. See 220a5a2841172a07d6d7849596316f0e0933e272.

0xB10C commented:
> demo scripts in contrib/tracing: right now, bpftrace scripts work, bcc scripts don't -- (https://github.com/bitcoin/bitcoin/pull/26593#issuecomment-1346186050)

Rebased and added a commit that fixes the bcc example scripts by using the PID of bitcoind instead of the file path. We already do this in the tests (https://github.com/bitcoin/bitcoin/commit/220a5a2841172a07d6d7849596316f0e0933e272).

willcl-ark reviewed (approved):
Lightly tested ACK.

1. I didn't manage to get `bpftrace` itself running on my Ubuntu system [yet](https://hackmd.io/@willcl-ark/SyZgMRbg3), but I was able to get the python scripts `log_raw_p2p_msgs.py`, `log_utxocache_flush.py` and `p2p_monitor.py` (using `bcc`) to run successfully under `sudo`.  
  (I do wonder about the viability of the `bpftrace` scripts in this project, seeing as they seem to be so brittle/difficult to use.)
2. The usdt tests all passed successfully:
  ```log
  interface_usdt_coinselection.py                        | ✓ Passed  | 5 s
  interface_usdt_net.py                                  | ✓ Passed  | 4 s
  interface_usdt_utxocache.py                            | ✓ Passed  | 11 s
  interface_usdt_validation.py                           | ✓ Passed  | 3 s
  ```
3. The new variadic macro is much simpler to reason about and I agree seems like a nice win all round. 
4. I was also able to see the semaphores being loaded from memory, compared to zero, and conditionally jumped to, as detailed in the PR description, e.g. for `CCoinsViewCache::AddCoin()`
  ```gdb
     ...
     0x0000000000b60927 <+615>:   movzwl 0xa64bea(%rip),%eax        # 0x15c5518 <utxocache_add_semaphore>
     0x0000000000b6092e <+622>:   cmp    $0x0,%eax
     0x0000000000b60931 <+625>:   jle    0xb609d6 <_ZN15CCoinsViewCache7AddCoinERK9COutPointO4Coinb+790>
     ...
  ```
5. I don't think having these semaphores in the global namespace should be a big issue for debugging or namespacing. So really it comes down to organisational priorities, where I'd agree with you that having them in the same file as the tracepoint feels cleaner.

0xB10C commented on src/test/util_trace_tests.cpp:
```diff
@@ -0,0 +1,32 @@
+  TRACEPOINT(test, one_arg, 1);
+  TRACEPOINT(test, six_args, 1, 2, 3, 4, 5, 6);
+  TRACEPOINT(test, twelve_args, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12);
+  BOOST_CHECK(true);
```
Yes, from the OP: "Also adding unit tests that include all tracepoint macros to make sure there are no compiler problems with them (e.g. some varadiac extension not supported)." I added them after addressing https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1034160141

But I think it makes sense to document this in the unit tests with a comment too. Will add on next push.

martinus commented on src/util/trace.h:
```diff
@@ -7,37 +7,26 @@
-#define TRACE11(context, event, a, b, c, d, e, f, g, h, i, j, k) DTRACE_PROBE11(context, event, a, b, c, d, e, f, g, h, i, j, k)
-#define TRACE12(context, event, a, b, c, d, e, f, g, h, i, j, k, l) DTRACE_PROBE12(context, event, a, b, c, d, e, f, g, h, i, j, k, l)
+// A USDT tracepoint with no arguments.
+#define TRACEPOINT0(context, event) \
```
Macro magic to the rescue! This works for me:

```patch
diff --git a/src/test/util_trace_tests.cpp b/src/test/util_trace_tests.cpp
index 6e3e846b4f..952b9cc6e4 100644
--- a/src/test/util_trace_tests.cpp
+++ b/src/test/util_trace_tests.cpp
@@ -18,7 +18,7 @@ BOOST_FIXTURE_TEST_SUITE(util_trace_tests, BasicTestingSetup)
 
 BOOST_AUTO_TEST_CASE(test_tracepoint_zero_args)
 {
-  TRACEPOINT0(test, zero_args);
+  TRACEPOINT(test, zero_args);
   BOOST_CHECK(true);
 }
 
diff --git a/src/util/trace.h b/src/util/trace.h
index 816d8c295d..3a0aed391b 100644
--- a/src/util/trace.h
+++ b/src/util/trace.h
@@ -29,25 +29,19 @@
     unsigned short context##_##event##_semaphore __attribute__((section (".probes")))
 
 // Returns true if something is attached to the tracepoint.
-#define TRACEPOINT_ACTIVE(context, event) context##_##event##_semaphore > 0
-
-// A USDT tracepoint with no arguments.
-#define TRACEPOINT0(context, event) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBE(context, event)
+#define TRACEPOINT_ACTIVE(context, event, ...) context##_##event##_semaphore > 0
 
 // A USDT tracepoint with one to twelve arguments. It's checked that the
 // tracepoint is active before preparing its arguments.
-#define TRACEPOINT(context, event, ...) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBEV(context, event, __VA_ARGS__)
+#define TRACEPOINT(context, ...) \
+    if (TRACEPOINT_ACTIVE(context, __VA_ARGS__)) \
+        STAP_PROBEV(context, __VA_ARGS__)
 
 #else
 
 #define TRACEPOINT_SEMAPHORE(context, event)
 #define TRACEPOINT_ACTIVE(context, event) false
-#define TRACEPOINT0(context, event)
-#define TRACEPOINT(context, event, ...)
+#define TRACEPOINT(context, ...)
 
 #endif // ENABLE_TRACING
 ```

The trick is to make `event` part of the variadic macro, so that there's always at least 1 argument in the variadic

martinus replied:
Based on @willcl-ark's comment https://github.com/bitcoin/bitcoin/issues/26916#issuecomment-1480997053 here is a version that works without any warnings in clang and gcc:

```patch
diff --git a/src/test/util_trace_tests.cpp b/src/test/util_trace_tests.cpp
index 6e3e846b4f..952b9cc6e4 100644
--- a/src/test/util_trace_tests.cpp
+++ b/src/test/util_trace_tests.cpp
@@ -18,7 +18,7 @@ BOOST_FIXTURE_TEST_SUITE(util_trace_tests, BasicTestingSetup)
 
 BOOST_AUTO_TEST_CASE(test_tracepoint_zero_args)
 {
-  TRACEPOINT0(test, zero_args);
+  TRACEPOINT(test, zero_args);
   BOOST_CHECK(true);
 }
 
diff --git a/src/util/trace.h b/src/util/trace.h
index 816d8c295d..aa88771b1e 100644
--- a/src/util/trace.h
+++ b/src/util/trace.h
@@ -26,28 +26,41 @@
 // upon attaching to the tracepoint and decremented when detaching. This needs
 // to be a global variable. It's placed in the '.probes' ELF section.
 #define TRACEPOINT_SEMAPHORE(context, event) \
-    unsigned short context##_##event##_semaphore __attribute__((section (".probes")))
+    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))
+
+// Extract the first argument of a variable number of arguments, even without warning
+// when only 1 argument is provided
+#define TRACEPOINT_FIRST_ARG(...) TRACEPOINT_FIRST_ARG_HELPER(__VA_ARGS__, dummy)
+#define TRACEPOINT_FIRST_ARG_HELPER(arg1, ...) arg1
 
 // Returns true if something is attached to the tracepoint.
-#define TRACEPOINT_ACTIVE(context, event) context##_##event##_semaphore > 0
+#define TRACEPOINT_ACTIVE(context, event) TRACEPOINT_ACTIVE_HELPER(context, event)
+#define TRACEPOINT_ACTIVE_HELPER(context, event) context##_##event##_semaphore > 0
 
-// A USDT tracepoint with no arguments.
-#define TRACEPOINT0(context, event) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBE(context, event)
+#if defined(__clang__)
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_PUSH _Pragma("clang diagnostic push") _Pragma("clang diagnostic ignored \"-Wgnu-zero-variadic-macro-arguments\"")
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_POP _Pragma("clang diagnostic pop")
+#else
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_PUSH
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_POP
+#endif
 
 // A USDT tracepoint with one to twelve arguments. It's checked that the
 // tracepoint is active before preparing its arguments.
-#define TRACEPOINT(context, event, ...) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBEV(context, event, __VA_ARGS__)
+#define TRACEPOINT(context, ...)                                             \
+    do {                                                                     \
+        if (TRACEPOINT_ACTIVE(context, TRACEPOINT_FIRST_ARG(__VA_ARGS__))) { \
+            TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_PUSH                       \
+            STAP_PROBEV(context, __VA_ARGS__);                               \
+            TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_POP                        \
+        }                                                                    \
+    } while (0)
 
 #else
 
 #define TRACEPOINT_SEMAPHORE(context, event)
 #define TRACEPOINT_ACTIVE(context, event) false
-#define TRACEPOINT0(context, event)
-#define TRACEPOINT(context, event, ...)
+#define TRACEPOINT(context, ...)
 
 #endif // ENABLE_TRACING
```

0xB10C replied:
Do you remember why you added a `do { ... } while (0)` to the `TRACEPOINT` macro?

kouloumos commented on src/util/trace.h:
```diff
@@ -7,39 +7,48 @@
+    unsigned short context##_##event##_semaphore __attribute__((section (".probes")))
+
+// Returns true if something is attached to the tracepoint.
+#define TRACEPOINT_ACTIVE(context, event) context##_##event##_semaphore > 0
```
I see that there is a `PASTE` macro that can be used instead of directly using the token-pasting operator (##), but I can't figure out why it exists and if it offers any advantage. 
https://github.com/bitcoin/bitcoin/blob/db720b5a703c90625fa7a4773bd2db5672427cbe/src/util/macros.h#L8

kouloumos reviewed (commented):
<details><summary> I've measured the number of executed instructions for each tracepoint based on the current and new implementation and verified that the overhead for tracepoints' arguments preparation - when enabled but not hooked - has been eliminated. Click to expand for details. </summary>
<br/>

master (49b87bfe7e2799d25ce709123ecafa872b36e87a)

Tracepoint|Disabled|Enabled (not hooked)|Enabled (hooked)
:---|:---|:---|:---
mempool:added|1|24|24
mempool:removed|1|136|136
mempool:replaced|4|60|60
mempool:rejected|2|250|250
net:inbound_message|1|50|50
net:outbound_message|3|43|43
validation:block_connected|1|8|8
utxocache:flush|2|95|95
utxocache:add|1|6|6
utxocache:spent|1|6|6
utxocache:uncache|1|6|6
coin_selection:selected_coins|4|107|107
coin_selection:normal_create_tx_internal|1|10|10
coin_selection:attempting_aps_create_tx|1|2|2
coin_selection:aps_create_tx_internal|-|11|11

PR (rebased as part of [26593-additions](https://github.com/kouloumos/bitcoin/tree/26593-additions))

Tracepoint|Disabled|Enabled (not hooked)|Enabled (hooked)
:---|:---|:---|:---
mempool:added|1|2|29
mempool:removed|1|7|146
mempool:replaced|4|3|60
mempool:rejected|2|3|254
net:inbound_message|1|2|54
net:outbound_message|3|3|52
validation:block_connected|1|3|10
utxocache:flush|2|2|98
utxocache:add|1|3|8
utxocache:spent|1|3|8
utxocache:uncache|1|3|9
coin_selection:selected_coins|4|3|109
coin_selection:normal_create_tx_internal|1|3|12
coin_selection:attempting_aps_create_tx|1|4|6
coin_selection:aps_create_tx_internal|-|3|12

The data are from my python gdb script that automatically measures executed instructions for all scenarios. You can find it in my [26593-additions](https://github.com/kouloumos/bitcoin/tree/26593-additions) branch alongside usage details in order to reproduce the results.
```
# extend_gdb.py
To make this file available to your gdb process:
1. Make sure that this file and `tracepoint_helper.py` are in the same dir
2. (gdb) source extend_gdb.py

You can now run the custom commands of this file.
The main command is:
(gdb) scenario run <scenario> <tracepoint> (--no-build)

Running it with no arguments will run all scenarios:
- tracepoints disabled
- tracepoints enabled
- tracepoints enabled (hooked)

The records are then saved in "output_<commit-hash>.txt"
```

Notes for the preliminary work on that branch:
- it's a rebased vesion of this PR in order to include the newly added mempool tracepoints.
- includes "missing" bpftrace scripts because these are what I'm using in my script to enable probes/hook on tracepoints.
</details>

 Regarding the `TRACEPOINT_SEMAPHORE(event, context)` macros: I don't see any advantage on having them in the same file. I consider them as part of tracepoints' inner workings which are already abstracted in another file. But still not opposing the current implementation as (based on my attempts) having them in `trace.h` is not possible and the inclusion of an extra file probably adds unnecessary complexity at this point. However, with my alternative approach (explained below) those are not needed.

#### macos support compatibility

I've rebased https://github.com/bitcoin/bitcoin/pull/25541 on top of this PR in [macos-usdt-support-26593](https://github.com/kouloumos/bitcoin/commits/macos-usdt-support-26593) and it works; there is only an issue with the new unit test. Because the macos implementation doesn't support arbitrary tracepoints, the `TRACEPOINT` macro expands to "undeclared identifier" error. To solve it, I made a change to conditionally include that test file to the build process. As this test case is only conditionally applicable (empty if not linux), maybe this can also apply to this PR.

#### Another approach(?)

This PR's gated approach reminded me of the [auto-generated header file](https://gist.github.com/kouloumos/df773b1e0896adca827e7960eace1d21) from `probes.d` in https://github.com/bitcoin/bitcoin/pull/25541. In https://github.com/bitcoin/bitcoin/pull/25541#issuecomment-1174858051 you mentioned how that file could also be generated on Linux with dtrace supplied by Systemtap, so I replicated this sempahor-gated functionality using the semaphores of the [linux auto-generated `probes.h` file](https://gist.github.com/0xB10C/655e506e9e53092f4becf412284f572a). The main advantage of this alternative approach is the unified build process for linux and macos as the current (pending) implementation for macos already uses the `probes.d` file. It then becomes trivial to also support tracepoints on macos. What do you think of this approach? This is essentially a mashup of your work, the macos support branch and [eklitzke's probes_sdt branch](https://github.com/eklitzke/bitcoin/tree/probes_sdt). You can find my changes at the [26593-another-approach](https://github.com/kouloumos/bitcoin/tree/26593-another-approach) branch ([diff](https://github.com/kouloumos/bitcoin/compare/26593-additions...26593-another-approach)) , feel free to pick whatever you like.

0xB10C commented:
Thanks for the helpful comments @willcl-ark, @martinus and @kouloumos! I plan to address these and update this PR with your suggestions. Currently, I'm prioritizing other projects during the feature freeze, so it might take a few more weeks.

Dezzj reviewed (approved):
Approved

0xB10C commented:
I plan to revisit this in the next two months. Feel free to close if I don't do it before the end of 2023.

0xB10C committed:
tracing: dedup TRACE macros & rename to TRACEPOINT

This deduplicates the TRACEx macros by using systemtaps STAP_PROBEV[0]
variadic macro instead of the DTrace compability DTRACE_PROBE[1] macros.
Bitcoin Core never had DTrace tracepoints, so we don't need to use the
drop-in replacement for it. As noted in pr25541[2], these macros aren't
compatibile with DTrace on macOS anyway.

This also renames the TRACEx macro to TRACEPOINT to clarify what the
macro does: inserting a tracepoint vs tracing (logging) something.

[0]: https://sourceware.org/git/?p=systemtap.git;a=blob;f=includes/sys/sdt.h;h=24d5e01c37805e55c36f7202e5d4e821b85167a1;hb=ecab2afea46099b4e7dfd551462689224afdbe3a#l407
[1]: https://sourceware.org/git/?p=systemtap.git;a=blob;f=includes/sys/sdt.h;h=24d5e01c37805e55c36f7202e5d4e821b85167a1;hb=ecab2afea46099b4e7dfd551462689224afdbe3a#l490
[2]: https://github.com/bitcoin/bitcoin/pull/25541/files#diff-553886c5f808e01e3452c7b21e879cc355da388ef7680bf310f6acb926d43266R30-R31

Co-authored-by: Martin Leitner-Ankerl <martin.ankerl@gmail.com>

0xB10C committed:
tracing: only prepare tracepoint args if attached

Before this commit, we would always prepare tracepoint arguments
regardless of the tracepoint being used or not. While we already made
sure not to include expensive arguments in our tracepoints, this
commit introduces gating to make sure the arguments are only prepared
if the tracepoints are actually used. This is a win-win improvement
to our tracing framework. For users not interested in tracing, the
overhead is reduced to a cheap 'greater than 0' compare. As the
semaphore-gating technique used here is available in bpftrace, bcc,
and libbpf, users interested in tracing don't have to change their
tracing scripts while profiting from potential future tracepoints
passing slightly more expensive arguments. An example are mempool
tracepoints that pass serialized transactions. We've avoided the
serialization in the past as it was too expensive.

Under the hood, the semaphore-gating works by placing a 2-byte
semaphore in the '.probes' ELF section. The address of the semaphore
is contained in the ELF note providing the tracepoint information
(`readelf -n ./src/bitcoind | grep NT_STAPSDT`). Tracing toolkits
like bpftrace, bcc, and libbpf increase the semaphore at the address
upon attaching to the tracepoint. We only prepare the arguments and
reach the tracepoint if the semaphore is greater than zero. The
semaphore is decreased when detaching from the tracepoint.

This also extends the "Adding a new tracepoint" documentation to
include information about the semaphores and updated step-by-step
instructions on how to add a new tracepoint.

0xB10C commented:
Rebased and addressed comments

- https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1155488768: @martinus I took parts from your suggestion and added you as a co-author for the first commit. Thanks for the macro-magic. There's now only the `TRACEPOINT()` macro that works for tracepoints with 0 to 12 args.
- https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1140555021: I clarified unit tests and added a new test for a manual `TRACEPOINT_ACTIVE()` macro.
- https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1161156029: updated the GDB `info probes` output in the docs


I also thought a bit about @kouloumos alternative approach that would allow us to use the tracepoints on Linux, macOS and potentially *BSD (not tested). This would require `dtrace` during build-time. I'm not sure this is something we want. If we explore in this direction, then it's probably better to do it in a separate PR.

As the first commit is an overall simplification and cleanup, it _might_ make also sense to split it off into a separate PR while we are exploring alternatives.

While this is rebased and _ready for review_ again, I'd prefer progress on #25832 while we're exploring alternatives to #26593. I'll leave this as work-in-progress until we're clearer on alternatives.

0xB10C commented:
Rebased. Now includes the tracepoint changes from https://github.com/bitcoin/bitcoin/pull/25273.

fanquake commented:
@0xB10C can you give a summary of the current state of these changes? It's draft, and from the looks of the most recent comment, that's because it's waiting on `support for macOS and FreeBSD based on this branch.`. Is that still correct/being worked on?

0xB10C commented:
Thanks for the poke. Undrafting.

I have a branch building on top of this PR adding support for no-dependency no-overhead-if-not-used macOS tracing in a working state and a concept for FreeBSD to do the same. Still some work to do before I can open a PR. This PR was a draft as I wanted to make sure it's compatible with both the macOS and FreeBSD changes. I think it's mostly compatible with the changes here.

Nonetheless, I would prefer review on https://github.com/bitcoin/bitcoin/pull/25832 first from reviewers interested in tracepoints.

laanwj commented on contrib/tracing/log_raw_p2p_msgs.py:
```diff
@@ -132,8 +132,9 @@ def print_message(event, inbound):
 
-def main(bitcoind_path):
-    bitcoind_with_usdts = USDT(path=str(bitcoind_path))
+def main(pid):
```
Attaching by pid is a good idea, even aside from semaphores, as it attaches a specific instance instead of all processes sharing the binary, which i think it does if you provide the path.

laanwj commented:
For some reason, after building with this, I dont get any output from:
```
$ gdb src/bitcoind
...
(gdb) info tracepoints
No tracepoints.
```
Tested with gdb `15.0.50.20240219-git` and `13.1`.

I checked that the tracepoints are included:
```
$ readelf -n ./src/bitcoind | grep NT_STAPSDT -A 4 -B 2
  stapsdt              0x00000062       NT_STAPSDT (SystemTap probe descriptors)
    Provider: net
    Name: outbound_message
    Location: 0x00000000001809bc, Base: 0x0000000000a2cb60, Semaphore: 0x0000000000b73248
    Arguments: -8@%r14 8@%r15 8@-560(%rbp) 8@24(%rcx) 8@%rax 8@%rdx
...
```
I don't know if this is an actual problem as we don't rely on gdb for tracing, but if so it's a documentation issue.

laanwj commented:
Oh, it's `info probes` not `info tracepoints`. Whoops. It checks out now.

Code review and lightly tested ACK 3933bdd6cfc9accab9e0ed47e83a5cc27ada6b68

maflcko reviewed (commented):
left a nit/question

0xB10C commented on src/util/trace.h:
```diff
@@ -11,39 +11,52 @@
+#define TRACEPOINT(context, ...)                                                \
+    do {                                                                        \
+        if (TRACEPOINT_ACTIVE(context, TRACEPOINT_FIRST_ARG(__VA_ARGS__))) {    \
+            STAP_PROBEV(context, __VA_ARGS__);                                  \
```
Tried this, but [didn't work for zero arguments tracepoint](https://github.com/bitcoin/bitcoin/actions/runs/8618313790/job/23620364926?pr=26593#step:6:4106) (the newly introduced tests catched this). However, we can use ` __VA_OPT__` from C++ 20.

0xB10C commented:
As mentioned in https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1558318881, with C++20, we can use `__VA_OPT__` and drop the changes from https://github.com/bitcoin/bitcoin/pull/26593#discussion_r1155488768. Done that and rebased.

0xB10C commented on src/util/trace.h:
```diff
@@ -11,39 +11,46 @@
+// automatically incremented by tracing frameworks (bpftrace, bcc, libbpf, ...)
+// upon attaching to the tracepoint and decremented when detaching. This needs
+// to be a global variable. It's placed in the '.probes' ELF section.
+#define TRACEPOINT_SEMAPHORE(context, event) \
+    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))
```
I remember trying a few combinations of this and in all cases the compiler/liker failed to find the semaphore. I just retried with the following patch which I think is similar to what you describe.

```patch
diff --git a/src/util/trace.h b/src/util/trace.h
index a4f0ea1ca4..8d7f6bdb74 100644
--- a/src/util/trace.h
+++ b/src/util/trace.h
@@ -27,8 +27,8 @@
 // automatically incremented by tracing frameworks (bpftrace, bcc, libbpf, ...)
 // upon attaching to the tracepoint and decremented when detaching. This needs
 // to be a global variable. It's placed in the '.probes' ELF section.
-#define TRACEPOINT_SEMAPHORE(context, event) \
-    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))
+#define TRACEPOINT_SEMAPHORE(context, event)
+//    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))

 #include <sys/sdt.h>

@@ -39,6 +39,7 @@
 // tracepoint is active before preparing its arguments.
 #define TRACEPOINT(context, event, ...)                                         \
     do {                                                                        \
+        static unsigned short context##_##event##_semaphore __attribute__((section(".probes"))); \
         if (TRACEPOINT_ACTIVE(context, event)) {                                \
             STAP_PROBEV(context, event __VA_OPT__(, ) __VA_ARGS__);                           \
         }                                                                       \

```

This (1) breaks the extra [`if(TRACEPOINT_ACTIVE(context, event))`](https://github.com/bitcoin/bitcoin/pull/26593/files#diff-86c34c3a3e5946d9f4bd754a5bf9c970b6a44a72c0ba216f1a2457a479ee9651R50) and (2) the linker complains it can't find the semaphores (e.g. `undefined reference to `utxocache_spent_semaphore'`). 

Otherwise, if we find a way to do it, I'd be happy to do it as I'm not too excited about having the semaphores as globals..

willcl-ark reviewed (approved):
ACK 38e24dfc5f1c54723018bf9d973b7d9cf9a2b824

I think the macro simplification included in this change is a nice win for simplicity.
Ran all the updated examples in contrib/tracing, and played around with adding/removing various tracepoint arguments to check everything works as expected.

Left one q and one suggestion inline, but neither a blocker for this.

0xB10C commented on doc/tracing.md:
````diff
@@ -270,21 +270,48 @@ the `util/trace.h` (defines the tracepoint macros) with `#include <util/trace.h>
+}
+```
+
+If needed, an extra `if(TRACEPOINT_ACTIVE(context, event) {..}` check can be
````
In some cases, tracepoint argument preparation can require more than a single readable line in the `TRACEPOINT` macro. An example is the transaction serialization [here](https://github.com/0xB10C/bitcoin/blob/6464cc95ca97c4535de79e52ed174e58ea0bfb18/src/txmempool.cpp#L467-L478). For context, see https://github.com/bitcoin/bitcoin/pull/26531#issuecomment-1333832906. I don't currently plan on PRing something like this, but had this running for more than a year to collect data on full-rbf RBF replacements.

```
    if(TRACEPOINT_ACTIVE(mempool, added)) {
        CDataStream added_tx(SER_NETWORK, PROTOCOL_VERSION);
        added_tx << entry.GetTx();
        TRACEPOINT(mempool, added,
            entry.GetTx().GetHash().data(),
            entry.GetTxSize(),
            entry.GetFee(),
            added_tx.size(),
            added_tx.data()
        );
    }
```

willcl-ark replied:
Ah I see, that makes sense now then, thanks.

vasild commented on contrib/tracing/mempool_monitor.py:
```diff
@@ -114,8 +114,9 @@
 
-def main(bitcoind_path):
-    bitcoind_with_usdts = USDT(path=str(bitcoind_path))
+def main(bitcoind_pid):
+    print(f"Hooking into bitcoind with pid {pid}")
+    bitcoind_with_usdts = USDT(pid=int(bitcoind_pid))
```
The argument is called `bitcoind_pid`, but the `print` uses `{pid}`. The neighboring `py` files use just `pid`:

```suggestion
def main(pid):
    print(f"Hooking into bitcoind with pid {pid}")
    bitcoind_with_usdts = USDT(pid=int(pid))
```

vasild reviewed (commented):
Approach ACK 38e24dfc5f1c54723018bf9d973b7d9cf9a2b824

Mostly nits that would be nice to address + one blocker about the `bitcoind_pid` vs `pid` mismatch.

0xB10C commented:
Addressed @vasild comments (thanks!) and rebased.

0xB10C commented on contrib/tracing/log_raw_p2p_msgs.py:
```diff
@@ -176,8 +177,8 @@ def handle_outbound(_, data, size):
 
 
 if __name__ == "__main__":
-    if len(sys.argv) < 2:
```
I decided against this. You don't strictly need to be root (https://hackmd.io/@willcl-ark/r19LVO2_6 or https://github.com/bitcoin/bitcoin/pull/24358#issuecomment-1083149220). Checking for getuid == 0 would require you to run as root.

vasild reviewed (approved):
ACK 1949b64afe7ecf91091f58623e75341a75996da0

Thank you!

0xB10C commented:
rebased (conflicted with https://github.com/bitcoin/bitcoin/pull/29071)

0xB10C commented:
rebased (conflicted with https://github.com/bitcoin/bitcoin/pull/30377)

0xB10C commented:
Rebased this on autotools removal, but the new way of detecting if the systemtap headers are present doesn't seem to work (I didn't do `cmake -B build -DWITH_USDT=ON` locally). Marking this as draft until I figure out how it works and rebase this on https://github.com/bitcoin/bitcoin/pull/30741

0xB10C commented:
rebased after #30741 and fixed the compilation with `-DWITH_USDT=ON` (by also setting `SDT_USE_VARIADIC` in `cmake/module/FindUSDT.cmake`)

vasild reviewed (approved):
ACK 5be6cc771faea00915a31240d9b5afdcdd4aa52f

0xB10C commented:
rebased after conflict with https://github.com/bitcoin/bitcoin/pull/30409

vasild reviewed (approved):
ACK 8a5b224fb11fdaa84265aa6d082d6f5e0ed97336

willcl-ark commented:
utACK 0de3e96e333090548a43e5e870c4cb8941d6baf1

Based on `git range-diff 38e24df...0de3e96e333090548a43e5e870c4cb8941d6baf1` .

The cmake tidy-up looks good to me, and I examined newly built binaries for probe and semaphore info. I didn't re-run the testing from my previous ACK.

vasild reviewed (approved):
ACK 0de3e96e333090548a43e5e870c4cb8941d6baf1

laanwj commented:
re-ACK 0de3e96e333090548a43e5e870c4cb8941d6baf1

jb55 commented:
utACK 0de3e96e333090548a43e5e870c4cb8941d6baf1

0xB10C commented:
Less overhead for everyone not hooking into the tracepoints :partying_face:. Now that this is merged, here are a few ideas I had for future tracepoint interface work. Noting them for prosperity. 
1. We _could_ internalize the relevant macro parts of systemtap's `sys/sdt.h` for the Linux tracepoints. This would allow us to drop the external dependency on systemtap, as we don't use 99% of it. Some prior commentary on this can be found here: https://github.com/hebasto/bitcoin/pull/162#issuecomment-2074645621
2. In the past I've managed to use a simple macro build a bitcoind with tracepoints on macOS. While our ebpf based demo scripts aren't compatible, @kouloumos DTrace scripts from https://github.com/bitcoin/bitcoin/pull/25541 are. This could look similar to https://github.com/0xB10C/bitcoin/blob/13b0ce221600fc7040502c834c51433ca96f91c3/src/util/trace.h#L35-L63. However, I currently don't have access to a macOS system to further work on this - I'm looking to rent one.
3. The same could possible on FreeBSD with e.g. these macros  https://github.com/0xB10C/bitcoin/blob/13b0ce221600fc7040502c834c51433ca96f91c3/src/util/trace.h#L104-L119.  I haven't tested this on FreeBSD yet. In https://github.com/bitcoin/bitcoin/pull/27458#pullrequestreview-1387810492, vasild mentiones he'd interested in FreeBSD tracepoints. My understanding is that the same macOS DTrace scripts from 25541 would work there too.
4. At least for macOS, we'd need an per-tracepoint  interface definition similar to https://github.com/0xB10C/bitcoin/blob/13b0ce221600fc7040502c834c51433ca96f91c3/src/util/trace.h#L121-L236. With some more commentary, these _could_ replace the list of tracepoints in https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#tracepoint-documentation. This would solve something similar to https://github.com/bitcoin/bitcoin/pull/29877#issuecomment-2061028339.
5. Even if we don't do 4. (because we e.g. don't want to do 2.), casting the tracepoint arguments to the type we expect to pass would be worthwhile to avoid problems like https://github.com/bitcoin/bitcoin/pull/29877. For some of our traceponts, we already do this: e.g. https://github.com/bitcoin/bitcoin/blob/900b17239fb25750fd30b4af6e6c38096374b71f/src/validation.cpp#L2902-L2907


Two other ideas that were mentioned in the past:

6. We *could* drop the example scripts from  `/contrib/tracing/*` and maintain them, along with proper tests in a CI, Bitcoin Core version compatibility information, possibly libbpf-based C or Rust tools (https://github.com/bitcoin/bitcoin/issues/30298), ... in an external repository like, for example, `0xb10c/tracing-scripts`,  `bitcoin-core/tracing-scripts`, or `bitcoin-monitoring/tracing-scripts` (what ever would works best).
7. If we at some point decide that maintaining the tracepoints in Bitcoin Core adds too much maintenance burden compared to the actual usage they're getting, we could drop the tracepoints but keep the tracepoint interface. We now have a unit test that includes a few nop tracepoints to check that the interface will still compile (https://github.com/0xB10C/bitcoin/blob/0de3e96e333090548a43e5e870c4cb8941d6baf1/src/test/util_trace_tests.cpp). This would allow us to drop the bcc python dependency in the CI and to remove the `interface_usdt_*` functional tests (which need to run in VM and can't run in a container). Tracepoint users could maintain a patch on Bitcoin Core adding the tracepoints (and tests) they want back in.
